From af2e9e3bb173885106f041f89b9f39b57fe3550a Mon Sep 17 00:00:00 2001
From: Jeremy Huddleston Sequoia <jeremyhu@apple.com>
Date: Mon, 11 Aug 2014 13:15:24 -0700
Subject: [PATCH 4/4] XQuartz: Better support turning off "Displays have
 separate Spaces" on OS X Mavericks

http://xquartz.macosforge.org/trac/ticket/1876

Follow-up to: 1c10b37380d228b35db8a8616a6312ac54f5e59b

Signed-off-by: Jeremy Huddleston Sequoia <jeremyhu@apple.com>
(cherry picked from commit ab32ee35890be51137b56525826ffe219b63eb88)
---
 hw/xquartz/xpr/xprScreen.c | 23 +++++++++++++++++------
 1 file changed, 17 insertions(+), 6 deletions(-)

diff --git xorg-server-1.16.0/hw/xquartz/xpr/xprScreen.c xorg-server-1.16.0/hw/xquartz/xpr/xprScreen.c
index 7aa1ae1..d0a525f 100644
--- xorg-server-1.16.0/hw/xquartz/xpr/xprScreen.c
+++ xorg-server-1.16.0/hw/xquartz/xpr/xprScreen.c
@@ -169,14 +169,25 @@ displayScreenBounds(CGDirectDisplayID id)
               (int)frame.size.width, (int)frame.size.height,
               (int)frame.origin.x, (int)frame.origin.y);
 
-    /* Remove menubar to help standard X11 window managers.
-     * On Mavericks and later, the menu bar is on all displays.
-     */
-    if (XQuartzIsRootless
+    Boolean spacePerDisplay = false;
 #if MAC_OS_X_VERSION_MIN_REQUIRED < 1090
-        && (NSAppKitVersionNumber >= 1265 || (frame.origin.x == 0 && frame.origin.y == 0))
+    if (NSAppKitVersionNumber >= 1265)
 #endif
-        ) {
+    {
+        Boolean ok;
+        (void)CFPreferencesAppSynchronize(CFSTR("com.apple.spaces"));
+        spacePerDisplay = ! CFPreferencesGetAppBooleanValue(CFSTR("spans-displays"),
+                                                            CFSTR("com.apple.spaces"),
+                                                            &ok);
+        if (!ok)
+            spacePerDisplay = true;
+    }
+
+    /* Remove menubar to help standard X11 window managers.
+     * On Mavericks and later, the menu bar is on all displays when spans-displays is false or unset.
+     */
+    if (XQuartzIsRootless &&
+        (spacePerDisplay || (frame.origin.x == 0 && frame.origin.y == 0))) {
         frame.origin.y += aquaMenuBarHeight;
         frame.size.height -= aquaMenuBarHeight;
     }
-- 
2.0.4

